### PR TITLE
[RFC] Scan state in hard forks

### PR DESCRIPTION
**Pull request by psteckler**
This RFC contains a proposal for handling the scan state across hard forks.

The suggested approach builds on the safe/unsafe distinction from #5019.

### CODE REVIEW

**Comment by mrmr1993** [comment]
For disasters, my intuition is that we should just bin the scan state and start over. I think we should try to meddle with the chain as little as possible between the pre-fork and post-fork state, since it's out of the view of concensus. Progressing the chain offline to add the work will also make it harder to have the verified fork point we discussed in the previous RFC.

From [Echo's RFC](https://github.com/CodaProtocol/coda/blob/rfc/hard-forks/rfcs/0024-hard-forks.md#parallel-scan-transitions):

> It would be easiest to deal with changes to the parallel scan state or the SNARKs within it if it were drained and restarted cleanly when necessary. This would simplify things a lot and mean that the merge SNARKs didn't have to be able to verify both old and new proofs. Below is a scheme to do this. The disadvantage is that it requires effectively pausing progress while this happens, and induces its own complexity.
>
> We add a new, special, transaction type, a no-op. These don't do anything but take up space in the scan, and in particular they don't ever require a base or merge SNARK to be built. Taking up space in the scan means that SNARKs for past transactions have to be bought to include them. Observe that if all the transactions in the scan state are no-ops then discarding it is safe.

This seems like a good strategy to explore for planned forks, although probably out of scope for the current disaster recovery proposals.

**Comment by deepthiskumar** [comment]
> For disasters, my intuition is that we should just bin the scan state and start over. I think we should try to meddle with the chain as little as possible between the pre-fork and post-fork state, since it's out of the view of concensus. Progressing the chain offline to add the work will also make it harder to have the verified fork point we discussed in the previous RFC.

In disaster recovery, If we discarded the scan state entirely, then the transactions in it applied to the staged ledger (which is part of the protocol state) should be undone (equivalent to discarding the staged ledger )as well since proof for it will never be generated. This is effectively changing the account balances from pre-fork to post-fork. These transactions are supposed to be finalized, so discarding it would be break the consensus? 
If we kept the scan state and therefore the transactions in it, the case was that we did not want to merge old transaction proofs with the new ones if transaction snark changed. 

In the planned fork scenario, I agree that we should try making minimal changes to the state

**Comment by psteckler** [comment]
> If we discarded the scan state entirely

Could there be an approach between discarding the scan state, and polishing it off?

We have RPCs that return scan states, for example.

**Comment by psteckler** [comment]
> These transactions are supposed to be finalized, so discarding it would be break the consensus?

They have proofs in the transaction SNARK, but they're not yet in the SNARKed ledger, right? Are they really finalized?

**Comment by psteckler** [comment]
For the safe case, another possibility is to bake the scan state into a new binary, and load it into memory, as if nothing happened. That way, we need not discard transactions and their proofs.

**Comment by bkase** [comment]
A few folks spoke with Zaki about this today -- summary:

Consider the three forms of economic activity: User-initiated transactions (user commands), Payouts from txn fees or snark fees (Fee transfers), and block rewards (coinbase fee transfers)

If we must throw away something, it's most important to keep User-commands. The others, if they must, can be reverted as long as this is communicated clearly to folks beforehand.

Weak subjectivity should be avoided if at all possible. This means we should not commit to any data on-chain that hasn't been included in a snark. Avoiding weak subjectivity is more important than downtime (on the order of 24hrs) esp. in the short term.

Additionally, requiring altruistic behavior on the network (such as some amount of snark proofs needing to be recreated without any fees attached) is acceptable as long as it's democratized (anyone on the network could do the altruistic work, even if O(1) ends up being the one doing it in the short-term).

**Comment by nholland94** [comment]
> A few folks spoke with Zaki about this today -- summary:
> 
> Consider the three forms of economic activity: User-initiated transactions (user commands), Payouts from txn fees or snark fees (Fee transfers), and block rewards (coinbase fee transfers)
> 
> If we must throw away something, it's most important to keep User-commands. The others, if they must, can be reverted as long as this is communicated clearly to folks beforehand.
> 
> Weak subjectivity should be avoided if at all possible. This means we should not commit to any data on-chain that hasn't been included in a snark. Avoiding weak subjectivity is more important than downtime (on the order of 24hrs) esp. in the short term.
> 
> Additionally, requiring altruistic behavior on the network (such as some amount of snark proofs needing to be recreated without any fees attached) is acceptable as long as it's democratized (anyone on the network could do the altruistic work, even if O(1) ends up being the one doing it in the short-term).

A few questions regarding this:

If we need to keep user commands, but also cannot commit to un-snarked data in the chain, that means that we need to walk backwards from the best chain at the time we are recovering from and collect all un-snarked txns to be put back into the mempool, right? We will need to discuss how to handle collisions between txns in the current mempool and txns we are unrolling back into the mempool for the restart. Also, we will need to discuss whether or not we just take the mempool of one of our nodes, or if we collect mempools from multiple distribute nodes throughout the network and need to coalesce them into a single mempool to redistribute for the network restart.

**Comment by bkase** [comment]
> If we need to keep user commands, but also cannot commit to un-snarked data in the chain, that means that we need to walk backwards from the best chain at the time we are recovering from and collect all un-snarked txns to be put back into the mempool, right? We will need to discuss how to handle collisions between txns in the current mempool and txns we are unrolling back into the mempool for the restart. Also, we will need to discuss whether or not we just take the mempool of one of our nodes, or if we collect mempools from multiple distribute nodes throughout the network and need to coalesce them into a single mempool to redistribute for the network restart.

I think we explicitly can't do this. If a transaction is "finalized" aka it's included in a block that we deem is final, even though it's not yet snarked, we must eventually snark it. So it must not be added back to the mempool. This is a hard requirement -- downtime and/or altruism is acceptable to make this possible.

**Comment by nholland94** [comment]
> > If we need to keep user commands, but also cannot commit to un-snarked data in the chain, that means that we need to walk backwards from the best chain at the time we are recovering from and collect all un-snarked txns to be put back into the mempool, right? We will need to discuss how to handle collisions between txns in the current mempool and txns we are unrolling back into the mempool for the restart. Also, we will need to discuss whether or not we just take the mempool of one of our nodes, or if we collect mempools from multiple distribute nodes throughout the network and need to coalesce them into a single mempool to redistribute for the network restart.
> 
> I think we explicitly can't do this. If a transaction is "finalized" aka it's included in a block that we deem is final, even though it's not yet snarked, we must eventually snark it. So it must not be added back to the mempool. This is a hard requirement -- downtime and/or altruism is acceptable to make this possible.

Yes, but I was referring to all the txns in blocks up to finalization. We have to look `k` blocks back in order to find the most recently finalized block, so that's `k` blocks of txns which are not in the mempool, and would otherwise be dropped by the network.

**Comment by bkase** [comment]
> Yes, but I was referring to all the txns in blocks up to finalization. We have to look k blocks back in order to find the most recently finalized block, so that's k blocks of txns which are not in the mempool, and would otherwise be dropped by the network.

Ah yes in the case of a disaster we have to pick a block that the network agrees on -- my gut says it's safe for this to be "bitcoin-level" finalized rather than true k-blocks-back and in this case user commands that aren't finalized (at the minimum at "bitcoin-level") are okay to drop, and I think it's nice if it were added back to mempools, but I don't see this as a release blocker.

**Comment by nholland94** [comment]
> Ah yes in the case of a disaster we have to pick a block that the network agrees on -- my gut says it's safe for this to be "bitcoin-level" finalized rather than true k-blocks-back and in this case user commands that aren't finalized (at the minimum at "bitcoin-level") are okay to drop, and I think it's nice if it were added back to mempools, but I don't see this as a release blocker.

I don't think "bitcoin-level" finalization really transfers in the world of Ouroboros. I would highly question doing so given that "weak subjectivity should be avoided if at all possible" is a goal.

**Comment by psteckler** [comment]
> > For the safe case, another possibility is to bake the scan state into a new binary, and load it into memory, as if nothing happened. That way, we need not discard transactions and their proofs.
> 
> The problem with this approach is we would have to coordinate 51% of honest stake to update their software very quickly after this binary is published. But maybe this is the best option.

N.B.: We'd be using the scan state/staged ledger associated with the root, so we'd still be tossing out those beyond the root to the best tip.

File: rfcs/0033-scan-state-hard-fork.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+already been applied to the staged ledger, we will have to discard the
+staged ledger at the root, and when the chain resumes, using the
+SNARKed ledger at the root as the staged ledger, and an empty scan state.
+
+Discarding the staged ledger means that we're discarding the finalized
+transactions reflected in that ledger. Also, the staged ledger hash
+at the pause point is no longer valid, and needs to be recomputed
+from the SNARKed ledger used as the staged ledger.
+
+Alternatively, we could retain the scan state and staged ledger, by
```
[the remark this answers is not in the document]
Comment (psteckler) [reply rfcs/0033-scan-state-hard-fork.md:48]: Yes.

**Comment by psteckler** [comment]
> Weak subjectivity should be avoided if at all possible.

Specifically, we should not consider the staged ledger as if it's been verified.

**Comment by psteckler** [comment]
> One other approach that I've heard thrown around is to produce both proofs in the scan state for a period of time before a fork

For true disasters, we won't know when we're going to need to fork.

**Comment by deepthiskumar** [comment]
> Yes, but I was referring to all the txns in blocks up to finalization. We have to look `k` blocks back in order to find the most recently finalized block, so that's `k` blocks of txns which are not in the mempool, and would otherwise be dropped by the network.

I think we should try to persist the transactions in the blocks that would get discarded and re-add it to the txn-pool post fork. At the very least in the case of a planned fork

**Comment by bkase** [comment]
Actually I still think it’s unacceptable to rollback all k blocks in the common case. That’s reverting too much activity. Here’s an alternative:

Everyone who is interested can publicly post their transition frontiers to some forum or GitHub issue. Union all the trees together (rooted at the point of finality which should be shared). Now this is the point where it’s not fully clear to me, but I suspect there’s a way we can take some common prefix of all the longish paths and use the tip of the common prefix instead of k as our finality point. Does anyone see a problem with this or more clearly see a way to pick that block?

**Comment by psteckler** [comment]
> Does anyone see a problem with this or more clearly see a way to pick that block?

@bkase I'm not sure it's reasonable to put this burden on operators, and it makes the fork result kind of unpredictable.

Instead, we could write an RPC to get all the TF breadcrumbs from the nodes reachable from the node we choose to get the root, and get the longest common prefix. Operators wouldn't have to get involved.

Already, the telemetry RPC returns a list of state hashes from the best tip breadcrumbs. The RPC would be similar.

**Comment by bkase** [comment]
> I'm not sure it's reasonable to put this burden on operators, and it makes the fork result kind of unpredictable.
> Instead, we could write an RPC to get all the TF breadcrumbs from the nodes reachable from the node we choose to get the root, and get the longest common prefix. Operators wouldn't have to get involved.

I agree we should automate it as much as possible, but I think for both a planned hard fork and disaster recovery scenario there will be manual involvement with operators (as we need to coordinate folks upgrading their software anyway!)

**Comment by psteckler** [comment]
> I agree we should automate it as much as possible, but I think for both a planned hard fork and disaster recovery scenario there will be manual involvement with operators (as we need to coordinate folks upgrading their software anyway!)

Yes, the upcoming "umbrella" RFC mentions notifying operators of new software.

**Comment by imeckler** [comment]
I endorse starting with the "unsafe" method, since we will have to do this anyway in the case of e.g., changing the hash function.

**Comment by deepthiskumar** [comment]
Regardless of what block we choose to fork off of, the scan state at the block will have transactions that haven't been proven entirely. In the disaster scenario, these transactions are finalized. In the planned fork scenario, we should not be throwing away the transactions simply by virtue of this being planned. So we need to keep the transactions in the scan state. In any of the fork scenarios, it is possible that the transaction snark has changed/patched rending the proofs that are in the scan state at the pause point invalid. So the proposal is to keep the transactions as is but throw away all the proofs. When the network resumes (either after a disaster or after the planned fork) the proofs for these are provided for free (by who?) and no new transactions are included until the debt is cleared.

**Comment by vanishreerao** [comment]
I have the following common comment for all the approaches stated. Let me know if you all agree. 
1. Scan state should not be discarded if there is no need to. (For example, the fork is just to push a software upgrade of using a different signature scheme.) In this case, the special block should only contain the software upgrade info and not prescribe any particular protocol state. Note we may have a protocol state that is "bad" from the consensus protocol's perspective. Meaning, the block should just instruct clients from what slot to use the new software. They should continue with whatever their view is, without us meddling with the consensus protocol (i.e., without us prescribing a protocol state through the special block)
2. If the scan state needs to be discarded, which is true in the cases of disasters, then we are forced to prescribe a protocol state. In this case, we should prescribe what the consensus protocol prescribes, which is the one wrt k blocks back. That being said, it is possible that a part of the "discarded" block production work and snark work will end up being redone after the rewind. If it's implementation-wise possible, it might be a good idea to have one's view of the scan state pre-fork still kept in memory to just pop if that work needs to be redone. 
3. I think an odd case that doesn't fit either scenario is when we would like to change things about the consensus protocol itself, such as, the number of slots in an epoch or the active slot coefficient f. I'll do a writeup for this (going on vacation next week, so will get this done a week after).

**Comment by deepthiskumar** [comment]
2. If the scan state needs to be discarded, which is true in the cases of disasters, then we are forced to prescribe a protocol state. In this case, we should prescribe what the consensus protocol prescribes, which is the one wrt k blocks back. That being said, it is possible that a part of the "discarded" block production work and snark work will end up being redone after the rewind. If it's implementation-wise possible, it might be a good idea to have one's view of the scan state pre-fork still kept in memory to just pop if that work needs to be redone.

@vanishreerao I think we don't want to discard the scan state even for this case i.e., scan state of the finalized block since the transactions in them are finalized. We would discard the proofs in it and generate new ones once the network is back up

**Comment by vanishreerao** [comment]
> 1. If the scan state needs to be discarded, which is true in the cases of disasters, then we are forced to prescribe a protocol state. In this case, we should prescribe what the consensus protocol prescribes, which is the one wrt k blocks back. That being said, it is possible that a part of the "discarded" block production work and snark work will end up being redone after the rewind. If it's implementation-wise possible, it might be a good idea to have one's view of the scan state pre-fork still kept in memory to just pop if that work needs to be redone.
> 
> @vanishreerao I think we don't want to discard the scan state even for this case i.e., scan state of the finalized block since the transactions in them are finalized. We would discard the proofs in it and generate new ones once the network is back up

I think we mean the same :). I might have not been as clear.

**Pull request merged**